Checkout: Decouple existingCardProcessor from checkout

sirbrillig opened the pull request:
#### Changes proposed in this Pull Request

This moves the composite checkout payment processor function `existingCardProcessor` to its own file and converts it to TypeScript. This is the first payment processor to be converted to TypeScript so it will set the stage for the rest to be converted as well by cleaning up the types and API. Notably it creates a `PaymentProcessorOptions` type which is the data passed into each processor.

This also refactors the processor a bit to decouple it from the 'wpcom' `@wordpress/data` Store so it can be used in other parts of calypso more easily. It does this mostly by fetching data it needs within the closure inside `CompositeCheckout`, making it possible for `existingCardProcessor` to rely only on its inputs.

The reason we are starting with this processor is that it can be used for "one-click checkout" (see #48132).

#### Testing instructions

- Make sure your browser devtools are recording network calls, and make sure that those calls are preserved after a redirect.
- Add a plan with a domain to your cart and visit checkout.
- Complete checkout using a saved credit card.
- Verify that the transaction completes as expected and that you are redirected as expected.
- Verify that the network request to the `/me/transactions` endpoint includes `domainDetails`.

sirbrillig committed:
WIP: Move existingCardProcessor to own file

sirbrillig committed:
Include transactionOptions with dataForProcessor

Eventually we can get rid of transactionOptions entirely.

sirbrillig committed:
Move transaction translate functions to translate-cart file

They are not types and belong outside the types file.

sirbrillig committed:
Add CardProcessorOptions type

sirbrillig committed:
Add stripeConfiguration to dataForProcessor

sirbrillig committed:
Make TransactionRequestWithLineItems more permissive for falsy values

sirbrillig committed:
Remove stripeConfiguration from ExistingCard payment method

sirbrillig committed:
Finish existingCardProcessor

sirbrillig committed:
Allow stripeConfiguration to be null in CardProcessorOptions

This should never happen in production but it's really hard to avoid it
in the type.

sirbrillig committed:
Apply ExistingCardProcessorData type to processor function call

sirbrillig committed:
Guard against missing stripeConfiguration in existingCardProcessor

sirbrillig committed:
Move getDomainDetails to own file

sirbrillig committed:
Allow siteId to be a number in TransactionRequestWithLineItems

sirbrillig committed:
Remove createExistingCardMethod from composite-checkout README

sirbrillig committed:
Move getPostalCode to own file

sirbrillig committed:
Stop passing total to existing-card payment processor

sirbrillig committed:
Add mergeIfObjects helper

sirbrillig committed:
Call existingCardProcessor with additional data already included

sirbrillig committed:
Make existingCardProcessor require all data as arguments

sirbrillig committed:
Make sure siteId is always a string in the existingCardProcessor

sirbrillig committed:
Change getDomainDetails to return undefined for missing data

sirbrillig committed:
Make siteId a string or undefined in existingCardProcessor

sirbrillig committed:
Disallow country being undefined in TransactionRequestWithLineItems

sirbrillig committed:
Ensure all required fields are checked in isValidTransactionData

sirbrillig committed:
Change mergeIfObjects to support an array of objects

sirbrillig committed:
Change usage of mergeIfObjects to array

nbloomf commented:
Starting with a site on a free plan, I added a personal plan to the cart and entered checkout from the plans page. My contact info is pre-filled. When I try to pay with a stored card I get a "Required purchase data is missing" calypso error notification.

Checking the network tab, there is not a call to the transactions endpoint; checking the code it looks like this error comes from the new `existing-card-processor` code this PR introduces. I dumped the `transactionData` checked at that point to the console, and here's what I see:

```
country: "US"
domainDetails: undefined
items: (2) [{…}, {…}]
name: "Mr bean"
paymentMethodToken: "stripe:(normal-looking-stripe-slug)"
paymentPartnerProcessorId: "stripe"
postalCode: "#####"
siteId: "###"
storedDetailsId: ###
subdivisionCode: "TX"
```

Looking at `isValidTransaction`, it looks like we're expecting a `paymentMethodType` parameter which is not set here, causing the error.

sirbrillig committed:
Remove paymentMethodType from transactionData type as it is added later

sirbrillig commented:
Thanks for noticing that! I added that at the last minute to comply with the `TransactionRequestWithLineItems` type, but I failed to notice that `submitExistingCardPayment` is where `paymentMethodType` gets added. Therefore, the type we were actually validating was `Omit< TransactionRequestWithLineItems, 'paymentMethodType' >`.

I've updated the types and removed the broken validator.

sirbrillig commented:
I'm trying one more thing with this PR, as it's related. I'd like to see if we can avoid storing the transaction response in the "wpcom" data store and instead use the existing and more standardized mechanism of `useTransactionStatus` that's part of `composite-checkout`. It's going to take a few tweaks, though.

sirbrillig commented:
Waiting on https://github.com/Automattic/wp-calypso/pull/48140

sirbrillig committed:
Remove saveTransactionResponseToWpcomStore from existingCardProcessor

sirbrillig committed:
Change mergeIfObjects to accept variadic args rather than an array

sirbrillig committed:
Change isValidTransactionData to throw specific errors

sirbrillig committed:
Rename CardProcessorOptions to PaymentProcessorOptions

sirbrillig commented:
This has been adjusted to work well with the upsell and is now ready for review again.

sirbrillig commented:
The failing tests are unrelated.

pottedmeat commented on client/my-sites/checkout/composite-checkout/types/transaction-endpoint.ts:
```diff
@@ -1,19 +1,44 @@
 
-const debug = debugFactory( 'calypso:composite-checkout:transaction-endpoint' );
+export interface TransactionRequestWithLineItems {
+	siteId: string | undefined;
```
There's a mixture of `prop?: string | undefined`, `prop?: string`, and `prop: string | undefined` that should probably be standardized.

sirbrillig replied:
I see what you mean. I think the different types are because `TransactionRequestWithLineItems` actually represents many different types. For each `paymentMethodType` there's a different set of required and optional fields, and this is just a catch-all.

I think you're right that `siteId` could probably be made optional safely since it can already be undefined. `paymentMethodToken`, on the other hand, should only ever be a string if it is set, which is why it was `paymentMethodToken?: string`, although you're right that as far as the data goes, they're the same.

Maybe I should try to create a different type for each transaction and make `TransactionRequestWithLineItems` a union of the specific types.

sirbrillig replied:
I pushed a couple commits to organize this type better and be more consistent.

sirbrillig committed:
Organize TransactionRequestWithLineItems properties

This also makes all optional props capable of holding `undefined`.

sirbrillig committed:
Clarify requried properties in ExistingCardTransactionRequest

sirbrillig committed:
Pass stripe/configuration directly to assignNewCardProcessor

sirbrillig commented:
Rebased after https://github.com/Automattic/wp-calypso/pull/48385

nbloomf commented:
Works as described; the transaction completes and I see `domainDetails` in the `me/transactions` call.

nbloomf reviewed (approved):
The test failures appear to be unrelated.
